### ISSUE TITLE
Black Ops 1 - Animated meshes - issue with unique objects / instancing (motionblur smear - motion vectors)

### ISSUE THREAD

**Issue by xoxor4d** [NVIDIAGameWorks/rtx-remix#450]
**Describe the bug**
- Meshes with animations (either skinned or rigid) appear very unstable with smearing and lighting issues in the game: Call of Duty - Black Ops (2010). Motion vectors can become very unstable (white flicker) if two or more of the same animated meshes are rendered. This is likely due to the game rendering the meshes in random order and remix not being able to decide which is which?

![image](https://github.com/NVIDIAGameWorks/rtx-remix/assets/45299104/ee772b1c-4366-4834-8c62-313039977268)

- Remix version used:
> dxvk-remix actions build: https://github.com/NVIDIAGameWorks/dxvk-remix/actions/runs/8437723042
> bridge-remix actions build: https://github.com/NVIDIAGameWorks/bridge-remix/actions/runs/8397501958

- Which GPU: rtx 3080
- Which GPU driver version: 551.76
- Which game build: Call of Duty - Black Ops 1 (2010) Singleplayer - Steam - (international build)

<br>

**To Reproduce**
1. Install CoD: Black Ops 1 (2011) Singleplayer - Steam (German version is NOT supported)
2. Install the v0.6 release of [T5-RTX](https://github.com/xoxor4d/t5-rtx/releases/tag/0.6) - my rtx-remix compatibility mod that re-implements fixed-function rendering - by dropping all the contents of the zip into the games root folder.
    > The v0.6 release comes with all required files including an ASI loader, the remix runtime files and a remix sample mod.

A. Directly load into the map `river` by using commandline arguments: `"path/BlackOps.exe" +set devmap river`
> OR  

B. Launch the game and then open the in-game console using `~` and use console command `devmap river`

<br>

**Expected behavior**
- clothing of npc's and other animated objects show visual artifacts (smearing and lighting artifacts)

> - Tweaking the `Unique Object Search Distance` does not seem to affect the behaviour at all (tried using extremely small and large values)
> - Assigning the `Motion Blur Mask-Out Texture` category to affected meshes does not get rid of the smearing effect
> - Disabling Motion blur entirely gets rid of the smear around meshes but meshes still have light artifacting (probably due to them rapidly changing positions in the scene)
> - Meshes are skinned on the cpu so hashes are very unstable

Video taken directly after loading the map `river`:
https://github.com/NVIDIAGameWorks/rtx-remix/assets/45299104/e8db4776-8d3a-4487-861e-561fe8566c3c

Video of an ape with only 1 draw call active (limited using dev settings):
https://github.com/NVIDIAGameWorks/rtx-remix/assets/45299104/c18f3f17-1e39-4819-8571-2472b19cac6c

More videos including motion vectors (ape01) and others:
https://drive.google.com/drive/folders/1Q7mb2yhKDSLjqK477-_3-7LMZgzmTArQ

Start of conversation about the issue with Mark on the showcase discord:
https://discord.com/channels/1028444667789967381/1152352915781058580/1221859907406663701

**Comment by AlexDunn** [comment]
Regarding this "This is likely due to the game rendering the meshes in random order and remix not being able to decide which is which?" - currently, Remix uses the "world" transform from fixed function API to determine the best matching instance frame-to-frame (simple radius based search - where radius is "Unique Object Search Distance").

Since the game is skinning these objects, do you know if they are baking the world transform into that skinning? (i.e. skinning in world space).  It's pretty common for games to do this.  In this case the world matrix being passed to Remix would be an identity, and it would explain this issue.

**Comment by xoxor4d** [comment]
Meshes are skinned in object space and transformed into the scene using the world matrix: https://github.com/xoxor4d/t5-rtx/blob/d05f63a92dc01d36102e655b8c608fc8c6b20e1c/src/components/modules/sp/fixed_function.cpp#L689

**Comment by AlexDunn** [comment]
And you verified that the world matrix is non-identity?  Looking at the code is not enough here.

**Comment by xoxor4d** [comment]
Oh true didn't think about that. Here is the world matrix for one of the apes:
![image](https://github.com/NVIDIAGameWorks/rtx-remix/assets/45299104/91cce687-df5c-47f6-9620-96cb8ac6adca)

What I've just noticed whilst doing that .. it seems like the game is rendering the (same?) mesh twice with really, really tiny offset and rotation differences. I mean like literally back to back .. so I guess that is what's causing that issue in the first place.

Hitting the same breakpoint again after the one above:
![Screenshot 2024-03-27 171432](https://github.com/NVIDIAGameWorks/rtx-remix/assets/45299104/498cea24-fe24-473d-9af7-ab169ddda78d)

I'll try to figure out why it does that, stop it from doing so and report back with the results

**Comment by AlexDunn** [comment]
Interesting.  It could be they are also using a very similar world transform, but the bulk of the transformation is happening in skinning.

Theoretically, this problem could also be solved on the Remix side.  Currently we use the world transform for instance matching, but this is not ideal - since game engines can bake any transform into the vertices.  Ideally we would use the bounding mesh centroid for our calculations as it would be much more reliable.  If you're interested in making a PR, I can walk you through the changes.  If not, it's something we'd like to fix one day for sure, just not sure where it lies on our roadmap.

**Comment by xoxor4d** [comment]
Okay I did some digging and it turns out that the function I was debugging is rendering some rigid, to me invisible, surface of the monkey (maybe a second special layer for the eyes or similar). The skinned monkey itself is rendered by another function and you are indeed correct, the transformation is baked into the vertices. The world matrix passed is identity with the 4th row being the origin of the player ... 

With bounding mesh centroid, you mean the center of the mesh's AABB right?

**Comment by xoxor4d** [comment]
Okay so I've tried using the bounding box center by passing RasterGeometry into `findSimilarInstance` and instead of using the transform `const Vector3 worldPosition = transform[3].xyz();`
I've changed it to `const Vector3 worldPosition = (geometry.boundingBox.minPos + geometry.boundingBox.maxPos) * 0.5f;`
(here: https://github.com/NVIDIAGameWorks/dxvk-remix/blob/7e24a527bd4d61cf238212ec853257446d973912/src/dxvk/rtx_render/rtx_instance_manager.cpp#L646)

That didn't help much so I've tried some hacky things on the games side.
I slightly changed the world matrix of each mesh by an offset that increased with the amount of rendered skinned meshes that frame (just to test if it would stop the smearing) but that didn't help either. 

I've now forced the game to only render the 33th skinned mesh which looks like that:

https://github.com/NVIDIAGameWorks/rtx-remix/assets/45299104/c67c5e90-df60-42d5-bcd1-3a3ed8b2c8bf

**Comment by xoxor4d** [comment]
I've somewhat fixed that issue from the game's side. The game normally skin's its actors in separate worker threads but that leads to a different rendering order each frame. There is still a little bit of flickering from time to time (white flashing when viewing motion vectors)

**Comment by AlexDunn** [comment]
There are a few more things we'd need for a Remix side solution.

Firstly, AABBs aren't always calculated - this is for perf reasons as in some cases its just extra unnecessary overhead.  Currently they are only calculated when anti-culling is enabled.  Otherwise they will just be 0,0,0.  So we should double check they are actually enabled in your case, and enable them if not.

Next there are some additional checks we'd need to modify to include the centroid here: https://github.com/NVIDIAGameWorks/dxvk-remix/blob/7e24a527bd4d61cf238212ec853257446d973912/src/dxvk/rtx_render/rtx_instance_manager.cp#L662

Additionally there is a spatial cache used to improve performance of instance searching.  Currently it uses the world transform - but here we'd need to use the centroid also: https://github.com/NVIDIAGameWorks/dxvk-remix/blob/7e24a527bd4d61cf238212ec853257446d973912/src/dxvk/rtx_render/rtx_instance_manager.cpp#L172

And here: https://github.com/NVIDIAGameWorks/dxvk-remix/blob/7e24a527bd4d61cf238212ec853257446d973912/src/dxvk/rtx_render/rtx_instance_manager.cpp#L183

There may be some other cases in instance manager that need to be covered I've missed here, but those are the important ones based on a quick read.

**Comment by xoxor4d** [comment]
Thanks for the write-up! I've made the following changes: https://github.com/xoxor4d/dxvk-remix/commit/e3bb0d11144ad24cbe5ef4fadcfd42ed89f869f9
(this does not include code to activate the AABB creation currently, I'm enabling anti culling for now)

Is usage of the blas bounding box correct? 
Would I also need to touch the blas creation here? https://github.com/NVIDIAGameWorks/dxvk-remix/blob/7e24a527bd4d61cf238212ec853257446d973912/src/dxvk/rtx_render/rtx_draw_call_cache.cpp#L88

I've not really noticed any difference in behavior or performance with these changes .. but CoD games might not be the best test case for that.

**Comment by AlexDunn** [comment]
Rather than adding the bounding box centroid to world transform position, we would need to do a matrix vector product.  This is so we factor scale and 

We can enable AABB calculations here, for now we could just return true:
 https://github.com/NVIDIAGameWorks/dxvk-remix/blob/7e24a527bd4d61cf238212ec853257446d973912/src/dxvk/rtx_render/rtx_options.cpp#L341

Lastly, with these changes in theory we should see more accurate instance matching.  Be sure to try changing the unique object search distance setting too.  Ideally you want to account for the maximum movement of a single object between frames.

This all assumes the game is splitting each individual mesh up for each drawcall.  If the game is batching all the monkeys up and rendering in a single drawcall, then it won't help.

**Comment by AlexDunn** [comment]
Yes looks like you would also need to make some changes here too, as you pointed out.
Good catch!!
 https://github.com/NVIDIAGameWorks/dxvk-remix/blob/7e24a527bd4d61cf238212ec853257446d973912/src/dxvk/rtx_render/rtx_draw_call_cache.cpp#L88

**Comment by xoxor4d** [comment]
Wow I think I'm getting somewhere with this! Thanks for all your guidance so far!
The little flickering that was left after the game sided fix is now gone 🙌

I'm kinda used to hlsl's `mul()` but it seems that there is no such function so I've created a translation matrix from the aabb centroid and multiplied that by the object's world matrix (would that be the correct way of doing it?) :
https://github.com/xoxor4d/dxvk-remix/commit/6df3434021e8dcd8e1b33a88fe03b1fc27b9d1cf

The change in the drawcall caching is similar: https://github.com/xoxor4d/dxvk-remix/commit/62033e43ee6c32b86d4b8a86aa9c900eb9c05e6b

**Comment by xoxor4d** [comment]
Alright the latest change includes an option (within the dev tab) to toggle my changes (also enables aabb creation) https://github.com/xoxor4d/dxvk-remix/commit/11dffa79359613ac29bbde7be27b8883ba115c03

Here is a video from cod4sp comparing the changes (on/off):
https://drive.google.com/file/d/15UkU3LXlrhrlwyqxXI3ULfyyf7GOfg2E

Definitely an improvement. Should I create a PR or is there something that I should change or add?

**Comment by AlexDunn** [comment]
Looking good!  Glad to hear that is helping.  If you're still seeing flickering sometimes, could it be that we just need to fine tune the unique object search distance?  Or could it be that the flickering is due to LOD transition?  In such cases we cant transfer the motion vectors reliably from one LOD to another.

Regarding the matrix vector product, you can just express it like so:
`Vector3 TransformedPos = (TransformMatrix * Vector4(Position.x, Position.y, Position.z, 1.f)).xyz();`

Other than that, I would probably move the new option to the Game Setup tab under "Heuristics".  Seems like this could be useful in many more places than just Black Ops!! 

For a nitpick: we could expose a getter on AxisAlignedBoundingBox which returns the centroid, rather than calculating it in place.  That would reduce some code duplication.

Lastly, this is amazing, thank you for doing this great work!  Please make an MR.  Would love to have @MarkEHenderson take a look too as he's recently familiar with this part of the code.

**Comment by xoxor4d** [comment]
Thanks again for your suggestions! I've implemented the getter method you've mentioned and cleaned up the transforms. I've also put the toggleable setting into the heuristics tab.

The leftover flickering is definitely due to LOD switching, but these changes seem to help with that as well (see the video I've linked in the PR). Reducing the Unique Search Distance didn't help much if at all (also visible in the mentioned video)
